src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -0,0 +1,1144 @@
+
+        EXAMPLES::
+
+            sage: F1 = IntegerValuedPolynomialRing(QQ)
+            sage: F2 = IntegerValuedPolynomialRing(QQ)
+            sage: F1 is F2
+            True
+        """
+        return super().__classcall__(cls, R)
+
+    def __init__(self, R):
+        r"""
+        Initialize ``self``.
+
+        INPUT:
+
+        - `R` -- a commutative ring
+
+        EXAMPLES::
+
+            sage: F = IntegerValuedPolynomialRing(QQ); F
+            Integer-Valued Polynomial Ring over Rational Field
+            sage: TestSuite(F).run()
+
+        TESTS::
+
+            sage: IntegerValuedPolynomialRing(24)
+            Traceback (most recent call last):
+            ...
+            TypeError: argument R must be a ring
+        """
+        if R not in Rings():
+            raise TypeError("argument R must be a ring")
+        cat = AlgebrasWithBasis(R).Commutative()
+        CombinatorialFreeModule.__init__(self, R, NonNegativeIntegers(),
+                                         latex_prefix="",
+                                         category=cat)
+
+    def _repr_(self) -> str:
+        r"""
+        Text representation of this integer-valued polynomial ring.
+
+        EXAMPLES::
+
+            sage: F = IntegerValuedPolynomialRing(QQ)
+            sage: F  # indirect doctest
+            Integer-Valued Polynomial Ring over Rational Field
+
+            sage: IntegerValuedPolynomialRing(ZZ)
+            Integer-Valued Polynomial Ring over Integer Ring
+        """
+        br = self.base_ring()
+        return f"Integer-Valued Polynomial Ring over {br}"
+
+    @cached_method
+    def one_basis(self):
+        r"""
+        Return the number 0, which index of `1` of this algebra,
+        as per :meth:`AlgebrasWithBasis.ParentMethods.one_basis`.
+
+        EXAMPLES::
+
+            sage: A = IntegerValuedPolynomialRing(QQ)
+            sage: A.one_basis()
+            0
+            sage: A.one()
+            B[0]
+        """
+        return self.basis().keys()(0)
+
+    def _an_element_(self):
+        """
+        Return an element of ``self``.
+
+        EXAMPLES::
+
+            sage: A = IntegerValuedPolynomialRing(QQ)
+            sage: A.an_element()
+            B[2] + 3*B[3]
+        """
+        B = self.basis()
+        return B[2] + 3 * B[3]
+
+    # mettre ici @cached_method ?
+    def product_on_basis(self, n1, n2):
+        r"""
+        Return the product of basis elements ``n1`` and ``n2``, as per
+        :meth:`AlgebrasWithBasis.ParentMethods.product_on_basis()`.
+
+        INPUT:
+
+        - ``n1``, ``n2`` -- integers
+
+        EXAMPLES::
+
+            sage: A = IntegerValuedPolynomialRing(QQ)
+            sage: A.product_on_basis(0, 1)
+            B[1]
+        """
+        i = ZZ(n1)
+        j = ZZ(n2)
+        if j < i:
+            j, i = i, j
+
+        return self.sum((-1)**k * i.binomial(k) * (i + j - k).binomial(i) *
+                        self.basis()[i + j - k] for k in range(i + 1))
```

Comment:
I believe all of the terms of distinct since i < j. Thus, this can be simplified to
```suggestion
        R = self.base_ring()
        return self._from_dict({i+j-k: R((-1)**k * i.binomial(k) * (i + j - k).binomial(i)) for k in range(i + 1)})
```
If I am wrong, then make it `self.linear_combination()` to avoid the transient elements.

It might also be good to have a more detailed doctest here.